File: RealEstateIQ/ml-service/app/schemas/prediction.py
"""
Pydantic schemas for the RealEstateIQ ML service.
"""

from typing import Dict, Optional
from pydantic import BaseModel, Field, field_validator

VALID_LOCATIONS = {"Colombo", "Kandy", "Galle", "Negombo"}


class PredictRequest(BaseModel):
    """Input schema for property price prediction."""

    area: float = Field(..., gt=0, le=50000, description="Area in square feet")
    bedrooms: int = Field(..., ge=1, le=20, description="Number of bedrooms")
    bathrooms: int = Field(..., ge=1, le=20, description="Number of bathrooms")
    location: str = Field(
        ..., description="Location: Colombo, Kandy, Galle, or Negombo"
    )
    house_age: int = Field(..., ge=0, le=150, description="Age of the house in years")
    parking: int = Field(..., ge=0, le=20, description="Number of parking spaces")

    @field_validator("location")
    @classmethod
    def validate_location(cls, v: str) -> str:
        if v not in VALID_LOCATIONS:
            raise ValueError(
                f"Invalid location '{v}'. Must be one of: {sorted(VALID_LOCATIONS)}"
            )
        return v

    model_config = {
        "json_schema_extra": {
            "example": {
                "area": 2200,
                "bedrooms": 4,
                "bathrooms": 3,
                "location": "Colombo",
                "house_age": 3,
                "parking": 2,
            }
        }
    }


class PredictResponse(BaseModel):
    """Prediction result schema."""

    predicted_price: float = Field(..., description="ML estimated property price (LKR)")
    price_per_sqft: Optional[float] = Field(
        None, description="Estimated price per square foot (LKR)"
    )
    model_version: str = Field(..., description="Model version identifier")
    algorithm: str = Field(..., description="Algorithm used")
    dataset_version: str = Field(..., description="Dataset version used for training")
    feature_importance: Dict[str, float] = Field(
        default_factory=dict, description="Feature importance scores from the model"
    )
    disclaimer: str = Field(
        default=(
            "This is an ML model estimate based on synthetic training data. "
            "It is not a guaranteed market valuation."
        )
    )


class HealthResponse(BaseModel):
    """Health check response schema."""

    status: str
    model_version: str
    algorithm: str
    model_loaded: bool


class ModelInfoResponse(BaseModel):
    """Detailed model information."""

    model_version: str
    algorithm: str
    metrics: Dict[str, float]
    dataset_version: str
    feature_importance: Dict[str, float]
    status: str

File: RealEstateIQ/backend/src/models/Prediction.ts
import mongoose, { Schema, Document } from 'mongoose';

export interface IPrediction extends Document {
  userId: mongoose.Types.ObjectId;
  propertyId?: mongoose.Types.ObjectId;
  inputFeatures: {
    area: number;
    bedrooms: number;
    bathrooms: number;
    location: string;
    house_age: number;
    parking: number;
  };
  predictedPrice: number;
  pricePerSqft: number | null;
  modelVersion: string;
  algorithm: string;
  datasetVersion: string;
  featureImportance: Record<string, number>;
  createdAt: Date;
}

const predictionSchema = new Schema<IPrediction>(
  {
    userId: {
      type: Schema.Types.ObjectId,
      ref: 'User',
      required: true,
    },
    propertyId: {
      type: Schema.Types.ObjectId,
      ref: 'Property',
      default: null,
    },
    inputFeatures: {
      area: { type: Number, required: true },
      bedrooms: { type: Number, required: true },
      bathrooms: { type: Number, required: true },
      location: { type: String, required: true },
      house_age: { type: Number, required: true },
      parking: { type: Number, required: true },
    },
    predictedPrice: { type: Number, required: true },
    pricePerSqft: { type: Number, default: null },
    modelVersion: { type: String, required: true },
    algorithm: { type: String, required: true },
    datasetVersion: { type: String, required: true },
    featureImportance: { type: Map, of: Number, default: {} },
  },
  { timestamps: true }
);

predictionSchema.index({ userId: 1, createdAt: -1 });
predictionSchema.index({ createdAt: -1 });

export const Prediction = mongoose.model<IPrediction>('Prediction', predictionSchema);
